concordium-std-derive/src/lib.rs

```diff
@@ -1053,65 +1084,308 @@ fn contract_function_schema_tokens(
+                            error,
+                            event: schema::EventSchema {event_schema, event_tags: Some(#event_tags_ty.to_vec())}
+                        });
+                    })
+                }
+                _ => Some(quote! {
+                    let parameter = <#parameter_ty as schema::SchemaType>::get_type();
+                    let error = <#error_ty as schema::SchemaType>::get_type();
+                    let event_schema = <#event_ty as schema::SchemaType>::get_type();
+                    let schema_bytes = concordium_std::to_bytes(&schema::FunctionV3::ParamErrorEvent {
+                        parameter,
+                        error,
+                        event: schema::EventSchema {event_schema, event_tags: None}
+                    });
+                }),
+            }
+        }
+        (None, Some(return_value_ty), Some(error_ty), Some(event_ty)) => {
+            let return_value_ty = return_value_ty.parse::<syn::Type>()?;
+            let error_ty = error_ty.parse::<syn::Type>()?;
+            let event_ty = event_ty.parse::<syn::Type>()?;
+
+            match event_tags_option {
+                Some(event_tags) => {
+                    let event_tags_ty = event_tags.parse::<syn::Type>()?;
+
+                    Some(quote! {
+                        let return_value = <#return_value_ty as schema::SchemaType>::get_type();
+                        let error = <#error_ty as schema::SchemaType>::get_type();
+                        let event_schema = <#event_ty as schema::SchemaType>::get_type();
+                        let schema_bytes = concordium_std::to_bytes(&schema::FunctionV3::RvErrorEvent {
+                            return_value,
+                            error,
+                            event: schema::EventSchema {event_schema, event_tags: Some(#event_tags_ty.to_vec())}
+                        });
+                    })
+                }
+                _ => Some(quote! {
+                    let return_value = <#return_value_ty as schema::SchemaType>::get_type();
+                    let error = <#error_ty as schema::SchemaType>::get_type();
+                    let event_schema = <#event_ty as schema::SchemaType>::get_type();
+                    let schema_bytes = concordium_std::to_bytes(&schema::FunctionV3::RvErrorEvent {
+                        return_value,
+                        error,
+                        event: schema::EventSchema {event_schema, event_tags: None}
+                    });
+                }),
+            }
+        }
+        (Some(parameter_ty), Some(return_value_ty), Some(error_ty), Some(event_ty)) => {
+            let parameter_ty = parameter_ty.parse::<syn::Type>()?;
+            let return_value_ty = return_value_ty.parse::<syn::Type>()?;
+            let error_ty = error_ty.parse::<syn::Type>()?;
+            let event_ty = event_ty.parse::<syn::Type>()?;
+
+            match event_tags_option {
```

Comment:
I think if you refactor the type to be a `struct` with `Option` values this will become much less code.